### PR TITLE
Multi stage mineable block construction

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Added GetNewBlock Base node service request and modified the NewBlockTemplate response to return a block template.
- Block and header templates were introduced that can be used to construct new partial blocks.
- When a miner submits a block template, the base node will update the header with the correct MMR roots and return the block to the miner.
- Added proto requests and responses for GetNewBlock and NewBlock. Proto conversions for NewBlockTemplate and NewBlockHeaderTemplate were also added..
- Fetch_future_mmr_root was added to Blockchain_db, Memory_db and Lmdb_db. This allows the MMR roots to be calculated using the provided change set without altering the MMRs of the current longest chain. This functionality is required to construct a valid new mineable block.
- The RewindMmr write operation of Memory_db and Lmdb_db was altered to allow step_back of 0, when 0 is provided then the MMRs are reset and the current uncommitted changes are discarded. 

## Motivation and Context
A multi stage mineable block construction process is required to allow the Base Node to construct the initial block with transactions, allow the miner to add the coinbase and finally allow the Base node to update the MMR roots in the header.

## How Has This Been Tested?
Added a test to demonstrate the mineable block creation process and other tests for testing the calculation of future MMR roots.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/blocks/new_block_template.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2018 The Tari Project
```
Comment (SWvheerden) [inline base_layer/core/src/blocks/new_block_template.rs]:
```suggestion
// Copyright 2019 The Tari Project
```

File: base_layer/core/src/blocks/new_blockheader_template.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2018 The Tari Project
```
Comment (SWvheerden) [inline base_layer/core/src/blocks/new_blockheader_template.rs]:
```suggestion
// Copyright 2019 The Tari Project
```

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 3]
```diff
@@ -773,6 +793,7 @@ where T: BlockchainBackend
         height: u64,
     ) -> Result<(), ChainStorageError>
     {
+        // TODO: Changes applied to the Kernel, UTXO and RP MMRs should also be reverted.
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: Is there a ticket for this?

Comment (neonknight64) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: Yes, #1052

File: base_layer/core/src/base_node/comms_interface/comms_response.rs [hunk 4]
```diff
@@ -36,5 +36,6 @@ pub enum NodeCommsResponse {
     TransactionOutputs(Vec<TransactionOutput>),
     HistoricalBlocks(Vec<HistoricalBlock>),
     MmrState(MutableMmrState),
-    NewBlockTemplate(Block),
+    NewBlockTemplate(NewBlockTemplate),
+    NewBlock(Block),
 }
```
Comment (SWvheerden) [inline base_layer/core/src/base_node/comms_interface/comms_response.rs]: We need to add a request for target difficulty as well
#1067

File: base_layer/core/src/blocks/blockheader.rs [hunk 5]
```diff
@@ -126,6 +130,21 @@ impl BlockHeader {
         }
     }
 
+    pub fn from_template(header_template: NewBlockHeaderTemplate) -> Self {
```
Comment (CjS77) [inline base_layer/core/src/blocks/blockheader.rs]: Consider using `impl From<BlockHeaderTemplate>` instead?

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 6]
```diff
@@ -86,6 +86,14 @@ pub trait BlockchainBackend: Send + Sync {
     fn fetch_mmr_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
     /// Returns only the MMR merkle root without the state of the roaring bitmap.
     fn fetch_mmr_only_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
+    /// Fetches the merklish root for the MMR tree identified by the key after the current additions and deletions have
+    /// temporarily been applied. Deletions of hashes from the MMR can only be applied for UTXOs.
+    fn fetch_future_mmr_root(
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: We're not really "fetching" anything here. 
Is there a name that more accurately describes the function of this method?

`calculate_mmr_root` perhaps? The parameters are a strong indicator that it's not the current root we'll be getting.
`calculate_mmr_root_after_changes` too wordy?

File: base_layer/core/src/blocks/blockheader.rs (hunk 5 above)
Comment (neonknight64) [reply base_layer/core/src/blocks/blockheader.rs]: Change applied.

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 6 above)
Comment (neonknight64) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: I agree the parameters provide a good indicator of its function, the function name was changed to `calculate_mmr_root`.
